### PR TITLE
promotions and partial pawn last rank segfault fix

### PR DESCRIPTION
**Pull request by ItsSypher**

### CODE REVIEW

File: src/piece.c [hunk 1]
```diff
@@ -262,23 +262,36 @@ pawn_valid(position pos, cell game[SIZE_STD][SIZE_STD], uint8_t * game_flags)
...
+		&& game[pos.rank + 1 * side_sign][pos.file].piece == NULL){
 		valid_moves[move_idx++] = coords_to_pos(pos.rank + 2 * side_sign, pos.file);
-	}
-	if (game[pos.rank + side_sign][pos.file + side_sign].piece != NULL &&
-	    game[pos.rank + side_sign][pos.file + side_sign].side ==
-	    opposite(game[pos.rank][pos.file].side)) {
+		}
+	if (game[pos.rank + side_sign][pos.file + side_sign].piece != NULL 
+		&& game[pos.rank + side_sign][pos.file + side_sign].side == opposite(game[pos.rank][pos.file].side)
+		&& pos.rank + side_sign >= 0
```
Comment (steewbsd) [inline src/piece.c:280]: Not needed, we already check if the target cell has a piece, so checking for out of bounds in this case is redundant as (ideally) no piece should be out of bounds in the first place

File: src/piece.c [hunk 2]
```diff
@@ -262,23 +262,36 @@ pawn_valid(position pos, cell game[SIZE_STD][SIZE_STD], uint8_t * game_flags)
...
-	}
+		}
 	*game_flags = check_if_king(valid_moves, move_idx, game, *game_flags);
 
-	if (game[pos.rank + side_sign][pos.file - side_sign].piece != NULL &&
-	    game[pos.rank + side_sign][pos.file - side_sign].side ==
-	    opposite(game[pos.rank][pos.file].side)) {
+	if (game[pos.rank + side_sign][pos.file - side_sign].piece != NULL 
+		&& game[pos.rank + side_sign][pos.file - side_sign].side == opposite(game[pos.rank][pos.file].side)
+		&& pos.rank + side_sign >= 0
```
Comment (steewbsd) [inline src/piece.c:290]: Same here

File: src/piece.c [hunk 3]
```diff
@@ -262,23 +262,36 @@ pawn_valid(position pos, cell game[SIZE_STD][SIZE_STD], uint8_t * game_flags)
 	int		side_sign = 1;
 	if (game[pos.rank][pos.file].side == BLACK)
 		side_sign = -1;
-	if (game[pos.rank + side_sign][pos.file].piece == NULL)
+	if (game[pos.rank + side_sign][pos.file].piece == NULL 
+		&& pos.rank + side_sign >= 0 
+		&& pos.file >= 0
```
Comment (steewbsd) [inline src/piece.c:268]: No need to check the file as pawn cannot move horizontally (only when capturing, see next comment)

File: src/board.c [hunk 4]
```diff
@@ -198,8 +198,26 @@ move_piece(position origin, position target, board * game, position * amoves)
 	int		valid = validate_move(origin, target, game, amoves);
 	if (valid == ERROR)
 		return ERROR;
-	target_cell.piece =
-		origin_cell.piece;	/* set the new position to point to
+		// check if promotion is availaile for pawn
+		if (origin_cell.piece->ident=='p'&&((origin_cell.side==BLACK&&target.rank==0)||(origin_cell.side==WHITE&&target.rank==7))){
+			char promo = 'q'; // change this to a piece decided by user unput in the tui
+			switch (promo) {
```
Comment (steewbsd) [inline src/board.c:106]: Good, but we already have a function for this: ident_to_piece() @ piece.h

File: src/piece.c (hunk 1 above)
Comment (ItsSypher) [reply src/piece.c:280]: got it thanks

File: src/piece.c (hunk 3 above)
Comment (ItsSypher) [reply src/piece.c:268]: understood thanks

**Pull request merged**